README.md
# hwordcloud
<!-- badges: start -->
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/czxa/hwordcloud?branch=master&svg=true)](https://ci.appveyor.com/project/czxa/hwordcloud) [![Travis build status](https://travis-ci.org/czxa/hwordcloud.svg?branch=master)](https://travis-ci.org/czxa/hwordcloud) [![Rdoc](http://www.rdocumentation.org/badges/version/hwordcloud)](http://www.rdocumentation.org/packages/hwordcloud) ![](http://cranlogs.r-pkg.org/badges/grand-total/hwordcloud?color=green) ![](http://cranlogs.r-pkg.org/badges/hwordcloud?color=green) ![](http://cranlogs.r-pkg.org/badges/last-week/hwordcloud?color=green) [![Launch binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/czxa/hwordcloud/master) [![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental) ![](https://visitor-count-badge.herokuapp.com/total.svg?repo_id=czxa.hwordcloud) [![Code Climate](https://codeclimate.com/github/czxa/hwordcloud/badges/gpa.svg)](https://codeclimate.com/github/dwyl/esta) [![HitCount](http://hits.dwyl.com/czxa/hwordcloud.svg)](http://hits.dwyl.com/czxa/hwordclud)
<!-- badges: end -->

## Introduction

The [htmlwidgets](https://github.com/ramnathv/htmlwidgets) package is very interesting and powerful. All the time, I want write my own html widgets, but I failed. This is my third try, finally I success! and the `hwordcloud` package come out. As I first try using `htmlwidgets`, I hope you will tolerate some of the flaws of this R package, and I will improve it in the future. Next, let's use `hwordcloud`!

## Installation

```r
devtools::install_github('czxa/hwordcloud')
# or just use git
devtools::install_git("https://github.com/czxa/hwordcloud.git")
```

## vignettes

[Rendering word clouds using R + Highcharts](https://czxb.github.io/br/hwordcloud.html)

## Licence

hwordclud has a dependency on Highcharts, a commercial JavaScript charting library. Highcharts offers both a commercial license as well as a free non-commercial license. Please review the licensing options and terms before using this software, as the hwordcloud license neither provides nor implies a license for Highcharts.

Highcharts (http://highcharts.com) is a Highsoft product which is not free for commercial and Governmental use.

## Acknowledgements

+ [jbkunst/highcharter](https://github.com/jbkunst/highcharter)
+ [Lchiffon/wordcloud2](https://github.com/Lchiffon/wordcloud2)

## Shiny Example

After Installation, you can run following codes to see hwordclud shiny app:

```r
dir <- system.file("examples", "hwordcloud", package = "hwordcloud")
setwd(dir)
shiny::shinyAppDir(".")
```

------------

<h4 align="center">

License

</h4>

<h6 align="center">

MIT © czxa.top

</h6>


NEWS.md
# hwordcloud 0.1.0

* Added a `NEWS.md` file to track changes to the package.


inst/examples/hwordcloud/app.R
#
# This is a Shiny web application. You can run the application by clicking
# the 'Run App' button above.
#
# Find out more about building applications with Shiny here:
#
#    http://shiny.rstudio.com/
#
if(!require(colourpicker)){
  install.packages("colourpicker", dependencies = T)
}
library(shiny)
library(colourpicker)
library(wordcloud2)

# Define UI for application that draws a histogram
ui <- fluidPage(

    # Application title
    titlePanel("Highcharter word cloud!"),

    # Sidebar with a slider input for number of bins
    sidebarLayout(
        sidebarPanel(
            selectInput(
                inputId = "theme",
                label = "Choose A Theme",
                choices = c("darkgreen",
                            "darkblue",
                            "avocado",
                            "darkunica",
                            "gray",
                            "gridlight",
                            "grid",
                            "sandsignika",
                            "sunset"),
                selected = "sandsignika"
            ),
            textInput(
                inputId = "title",
                label = "Input Chart Title: ",
                placeholder = "Word Cloud",
                value = "Word Cloud"
            ),
            selectInput(
                inputId = "titleAlign",
                label = "Title Alignment: ",
                choices = c("left", "center", "right"),
                selected = 'center'
            ),
            textInput(
                inputId = "titleSize",
                label = "Title Size: ",
                placeholder = "20px",
                value = "20px"
            ),
            colourInput(
                inputId = "titleColor",
                label = "Title Color: ",
                value = "#333333",
                showColour = "background",
                allowTransparent = TRUE
            ),
            textInput(
                inputId = "subtitle",
                label = "Input Subtitle ",
                placeholder = "https://www.czxa.top",
                value = ""
            ),
            selectInput(
                inputId = "subtitleAlign",
                label = "Subtitle Alignment: ",
                choices = c("left", "center", "right"),
                selected = "center"
            ),
            textInput(
                inputId = "subtitleSize",
                label = "Subtitle Size: ",
                placeholder = "10px",
                value = ""
            ),
            colourInput(
                inputId = "subtitleColor",
                label = "Subtitle Color: ",
                value = "#666666",
                showColour = "background",
                allowTransparent = TRUE
            ),
            textInput(
                inputId = "iterm",
                label = "Attribute in toolbox: ",
                placeholder = "count",
                value = "count"
            ),
            textInput(
                inputId = "height",
                label = "Hight: ",
                placeholder = "500px",
                value = "500px"
            ),
            textInput(
                inputId = "width",
                label = "Width: ",
                placeholder = "100%",
                value = "100%"
            )
        ),

        # Show a plot of the generated distribution
        mainPanel(
           hwordcloud::hwordcloudOutput("shinyhwordcloud", height = "500px")
        )
    )
)

# Define server logic required to draw a histogram
server <- function(input, output) {
    library(hwordcloud)
    library(wordcloud2)
    df <- demoFreq %>%
        head(50)
    output$shinyhwordcloud <- hwordcloud::renderHwordcloud({
        hwordcloud::hwordcloud(text = df$word, size = df$freq,
             theme = input$theme,
             itermName = input$iterm,
             width = input$width,
             height = input$height,
             title = input$title,
             titleAlign = input$titleAlign,
             titleSize = input$titleSize,
             titleColor = input$titleColor,
             subtitle = input$subtitle,
             subtitleAlign = input$subtitleAlign,
             subtitleSize = input$subtitleSize,
             subtitleColor = input$subtitleColor
             )
    })
}

# Run the application
shinyApp(ui = ui, server = server)


R/hwordcloud.R
#' @title hwordcloud: Rendering word clouds using R + Highcharts
#'
#' @description This function can create wordclouds by binding R and Highcharts.
#'
#' @import htmlwidgets
#'
#' @param text character vector;
#' @param size numeric vector;
#' @param width chart width, for example, "100\%";
#' @param height chart height, for example, "400px";
#' @param theme chart theme, you can use these themes:
#' darkgreen/darkblue/avocado/darkunica/gray/
#' gridlight/grid/sandsignika/sunset;
#' @param itermName attribute in tooltip;
#' @param title title;
#' @param titleAlign title alignment, left/center/right;
#' @param titleSize title size, like "20px";
#' @param titleColor title color, like "#333333";
#' @param subtitle subtitle;
#' @param subtitleAlign subtitle alignment, left/center/right;
#' @param subtitleSize subtitle size, like "16px";
#' @param subtitleColor subtitle color, like "#666666"
#'
#' @examples
#' library(hwordcloud)
#' library(wordcloud2)
#' df <- head(demoFreq, 50)
#' if (interactive()) {
#'   hwordcloud(text = df$word, size = df$freq)
#' }
#' @export
hwordcloud <- function(
  text,
  size,
  width = "100%",
  height = NULL,
  theme = "sandsignika",
  itermName = "count",
  title = "",
  titleAlign = "center",
  titleSize = "20px",
  titleColor = "#333333",
  subtitle = "",
  subtitleAlign = "center",
  subtitleSize = "",
  subtitleColor = "#666666"
) {
  x <- list(
    text = text,
    size = size,
    theme = theme,
    itermName = itermName,
    title = title,
    titleAlign = titleAlign,
    titleSize = titleSize,
    titleColor = titleColor,
    subtitle = subtitle,
    subtitleAlign = subtitleAlign,
    subtitleSize = subtitleSize,
    subtitleColor = subtitleColor
  )

  # create widget
  htmlwidgets::createWidget(
    name = "hwordcloud",
    x,
    width = width,
    height = height,
    package = "hwordcloud"
  )
}

#' Shiny bindings for hwordcloud
#'
#' Output and render functions for using hwordcloud within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a hwordcloud
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name hwordcloud-shiny
#'
#' @export
hwordcloudOutput <- function(outputId, width = "100%", height = "400px") {
  htmlwidgets::shinyWidgetOutput(outputId, "hwordcloud", width, height, package = "hwordcloud")
}

#' @rdname hwordcloud-shiny
#' @export
renderHwordcloud <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) {
    expr <- substitute(expr)
  } # force quoted
  htmlwidgets::shinyRenderWidget(expr, hwordcloudOutput, env, quoted = TRUE)
}

#' hwordcloud package
#' @description Rendering word clouds using R + Highcharts
#' @section \code{\link{hwordcloud}}: Rendering word clouds using R + Highcharts
#' @docType package
#' @name hwordcloud
NULL
